README.markdown
Jump, a bookmarking system for the bash shell.
==============================================

Introduction
------------

Jump is a tool that allows you to quickly change directories in the bash
shell using bookmarks. Thanks to Jump, you won't have to type those long
paths anymore. 
Jump was inspired by go-tool by ActiveState
(<http://code.google.com/p/go-tool/>).

Usage
-----

Say you often work in a directory with a path like
`/Users/giuseppe/Work/Projects/MyProject`. With Jump you can add a bookmark
to it:

    $ jump add myproject

From now on you can jump to your project just by typing:

    $ jump to myproject

You can even append subfolders to the bookmark name!

    $ jump to myproject/src/
    
You can take a look at all your bookmarks with the `list` command:

    $ jump list

To delete a bookmark you don't need anymore, use the `delete` command:

    $ jump delete myproject

Don't remember a command? Just type:

    $ jump help

Installation
------------

I'm not shipping compiled binaries at the moment, so you'll have to compile
Jump from the sources. That's not hard though:

    $ gcc -o jump-bin jump.c
    $ sudo cp jump-bin /usr/bin

In order to be able to change directories, Jump needs a shell driver you'll
have to add to your bash configuration file (e.g. `~/.bash_profile`):

    function jump {
        if [ "$1" == "to" ] && [ "$2" != "" ]; then 
            cd $(jump-bin $*)
        else
            jump-bin $*
        fi
    }

If you prefer a shorter command to type even less, just rename the function
with a shorter name (e.g. `function j {...}`).

&copy; 2010 Giuseppe Capizzi (<g.capizzi@gmail.com>)

Changelog
Mon Jun 29 16:57:36 CEST 2015  Flavio Castelli <flavio@castelli.name>

	Version 0.4.1:
		* prevent incompatibility with recent versions of terminal-table

Wed Jan 21 14:32:36 CET 2015  Flavio Castelli <flavio@castelli.name>

	Version 0.4.0:
	* Added "--path" flag: this makes possible new creative uses of
		jump like: "vim `jump -p rails-app/README`".

Mon Jan 12 12:25:15 CET 2015  Flavio Castelli <flavio@castelli.name>

	Version 0.3.3:
		* Fixed typo when loading bookmarks fails

Thu Feb 21 12:49:52 CET 2013  Flavio Castelli <flavio@castelli.name>

	Version 0.3.2:
		* complete even after the first children.

Wed Feb 06 09:08:54 CET 2013  Flavio Castelli <flavio@castelli.name>

	Version 0.3.1:
	* exit with a nonzero value for unknown bookmarks.

Wed Mar 07 11:15:01 CET 2012  Flavio Castelli <flavio@castelli.name>

	Version 0.3.0:
	* better output when listing bookmarks.
	* added --bash-integration and --zsh-integration help tasks.

Sat Aug 28 09:04:30 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* added support to zsh

Wed Aug 11 15:14:05 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* updated the documentation

Tue Aug 10 14:32:11 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* updated the usage banner

Tue Aug 10 10:25:26 CEST 2010  Flavio Castelli  <flavio@castelli.name>

	* First release


test/test_helper.rb
# This file is part of the jump project
#
# Copyright (C) 2010 Flavio Castelli <flavio@castelli.name>
# Copyright (C) 2010 Giuseppe Capizzi <gcapizzi@gmail.com>
#
# jump is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# jump is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Keep; if not, write to the
# Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

require 'rubygems'
require 'test/unit'
require 'fakefs/safe'
require 'yaml'


test/bookmarks_test.rb
# This file is part of the jump project
#
# Copyright (C) 2010 Flavio Castelli <flavio@castelli.name>
# Copyright (C) 2010 Giuseppe Capizzi <gcapizzi@gmail.com>
#
# jump is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# jump is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Keep; if not, write to the
# Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

require "test_helper"
require 'bookmarks'
require 'fileutils'

class BookmarksTest < Test::Unit::TestCase

  def setup
    @test_bookmarks = { "foo" => "/tmp/foo",
                        "bar" => "/tmp/bar",
                        "complex" => "#{File.expand_path '~'}/tmp/foo/bar/complex"}
    FakeFS do
      FileUtils.mkdir_p(File.expand_path('~'))
      File.open(Bookmarks::BOOKMARKS_PATH, 'w') do |file|
        file << YAML::dump(@test_bookmarks)
      end

      @bookmarks = Bookmarks.new
    end
  end

  def test_expand_path
    @test_bookmarks.each do |bookmark, expected_path|
      current_path = @bookmarks.expand_path bookmark
      assert_equal expected_path, current_path
    end

    # check absolute path
    assert_equal "/foo", @bookmarks.expand_path("/foo")

    # check multiple / handling
    assert_equal "/tmp/foo/1/2", @bookmarks.expand_path("foo/1/2")
  end

  def test_add
    # should add a new bookmark
    bookmark = "flavio_home"
    path = "/home/flavio"
    @bookmarks.add(path, bookmark)
    assert_equal path, @bookmarks.expand_path(bookmark)

    #should update a bookmark
    new_path = "/tmp"
    @bookmarks.add(new_path, bookmark)
    assert_equal new_path, @bookmarks.expand_path(bookmark)
  end

  def test_to_s
    output = <<EOF
---------------------------------
 Bookmark  Path                  
---------------------------------
 bar       /tmp/bar              
 complex   ~/tmp/foo/bar/complex 
 foo       /tmp/foo              
---------------------------------
EOF
    assert_equal output.chomp, @bookmarks.to_s
  end

  def test_save
    bookmark = "flavio_home"
    path = "/home/flavio"
    @bookmarks.add(path, bookmark)

    expected_bookmarks = @test_bookmarks.dup
    expected_bookmarks[bookmark] = path

    FakeFS do
      @bookmarks.save

      assert File.exists?(Bookmarks::BOOKMARKS_PATH)
      contents = YAML::load_file Bookmarks::BOOKMARKS_PATH
      assert_equal expected_bookmarks, contents
    end
  end
end



test/completion_test.rb
# This file is part of the jump project
#
# Copyright (C) 2010 Flavio Castelli <flavio@castelli.name>
# Copyright (C) 2010 Giuseppe Capizzi <gcapizzi@gmail.com>
#
# jump is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# jump is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Keep; if not, write to the
# Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

require "test_helper"
require 'bookmarks'
require 'fileutils'

class BookmarkCompletionTest < Test::Unit::TestCase
  
  def setup
    FakeFS do
      @bookmarks = Bookmarks.new
      @bookmarks.add("/home/flavio/templates", "templates")
      @bookmarks.add("/home/flavio/templates", "test")
      @bookmarks.add("/home/flavio/test/rails_app", "rails")
      @bookmarks.add("/home/flavio/test/another_rails_app", "rails2")

      FileUtils.mkdir_p "/home/flavio/templates/foo/bar"
      FileUtils.mkdir_p "/home/flavio/templates/baz"
      FileUtils.mkdir_p "/home/flavio/test/rails_app/log"
      FileUtils.mkdir_p "/home/flavio/test/rails_app/locale"
      FileUtils.mkdir_p "/home/flavio/test/rails_app/app/model"
      FileUtils.touch   "/home/flavio/test/rails_app/local_file"
      FileUtils.mkdir_p "/home/flavio/test/another_rails_app"
    end
  end

  def test_absolute_path
    FakeFS do
      assert_equal "/rails", @bookmarks.complete('/rails')
    end
  end

  def test_empty_completes_to_all
    FakeFS do
      assert_equal "rails rails2 templates test", @bookmarks.complete(nil)
      assert_equal "rails rails2 templates test", @bookmarks.complete('')
    end
  end

  def test_no_matches # => should return the same text
    FakeFS do
      assert_equal "foo", @bookmarks.complete("foo")
    end
  end

  def test_no_matches_with_suffix
    FakeFS do
      assert_equal "foo/meh/zzz", @bookmarks.complete("foo/meh/zzz")
    end
  end

  def test_prefix
    FakeFS do
      assert_equal "templates/ test/", @bookmarks.complete("te")
      assert_equal "rails/ rails2/", @bookmarks.complete("ra")
    end
  end

  def test_nonexisting_suffix
    FakeFS do
      # /home/flavio/templates/bar doesn't exist => should return the same text
      assert_equal "templates/bar", @bookmarks.complete("templates/bar")
      assert_equal "templates/bar/1/2",
                   @bookmarks.complete("templates/bar/1/2")
    end
  end

  def test_completes_children_after_separator
    FakeFS do
      assert_equal "rails/app/ rails/locale/ rails/log/",
                   @bookmarks.complete("rails/")
    end
  end

  def test_completes_suffix
    FakeFS do
      assert_equal "rails/locale/ rails/log/",
                   @bookmarks.complete("rails/lo")
    end
  end

  def test_completes_suffix_path
    FakeFS do
      assert_equal "rails/app/model/",
                   @bookmarks.complete("rails/app/")
    end
  end

end


jump.c
/*
 * Jump, a bookmarking system for the bash shell.
 * Copyright (c) 2010 Giuseppe Capizzi
 * mailto: g.capizzi@gmail.com
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NAME_SIZE      64
#define PATH_SIZE     256

char bookmarks_filename[] = ".jumprc";

typedef struct bookmark_struct bookmark;
struct bookmark_struct {
    char *name;
    char *path;
    bookmark *next;
};

bookmark *bookmarks = NULL;

void exit_with_error(const char *message) {
    printf("*** ERROR: %s\n", message);
    exit(1);
}

void print_usage() {
    printf("jump to [bookmark]    Jumps to the directory pointed by [bookmark]\n");
    printf("jump add [bookmark]   Saves the current directory in [bookmark]\n");
    printf("jump del [bookmark]   Deletes [bookmark]\n");
    printf("jump list             Prints the list of all saved bookmarks\n");
    printf("jump help             Displays this message\n");
}

// Returns the current working directory.
char *cwd() {
    char *working_dir;
    FILE *fp;

    working_dir = calloc(PATH_SIZE, sizeof(char));

    fp = popen("pwd", "r");
    if (fp == NULL) {
        exit_with_error("failed to run pwd.");
    }
    fgets(working_dir, PATH_SIZE, fp);
    pclose(fp);

    return working_dir;
}

// Returns the path of the bookmarks file in the user home directory.
char *get_bookmarks_filepath() {
    char *homedir = getenv("HOME");
    int filepath_len = strlen(homedir) + strlen(bookmarks_filename) + 1;

    char *bookmarks_filepath;
    bookmarks_filepath = calloc(filepath_len, sizeof(char));
    strcpy(bookmarks_filepath, homedir);
    strcat(bookmarks_filepath, "/");
    strcat(bookmarks_filepath, bookmarks_filename);
    
    return bookmarks_filepath;
}

// Adds a bookmark to the list.
bookmark *add_bookmark(char *name, char *path) {
    bookmark *b;

    // create bookmark node
    b = malloc(sizeof(bookmark));
    b->name = strdup(name);
    b->path = strdup(path);
    b->next = NULL;

    // add bookmark to the list
    if (bookmarks == NULL) {
        bookmarks = b;
    } else {
        bookmark *p = bookmarks;

        // find the last but one node
        while (p->next != NULL) {
            p = p->next;
        }

        // add the node to the list
        p->next = b;
    }

    return b;
}

// Finds a bookmark in the list.
bookmark *find_bookmark(char *name) {
    bookmark *p = bookmarks;

    while (p != NULL && strcmp(p->name, name) != 0) {
        p = p->next;
    }

    return p;
}

// Creates a new bookmark if it does not exist. If it does already exist,
// replaces its path with the new one.
bookmark *new_bookmark(char *name, char *path) {
    bookmark *p;

    if ((p = find_bookmark(name)) != NULL) {
        strcpy(p->path, path);
    } else {
        p = add_bookmark(name, path);
    }

    return p;
}

// Deletes a bookmark from the list.
void delete_bookmark(char *name) {
    bookmark *d;

    // the record is the first in the list
    if (bookmarks != NULL && strcmp(bookmarks->name, name) == 0) {
        d = bookmarks;
        bookmarks = bookmarks->next;
        free(d);
    } else {
        bookmark *p = bookmarks;
        
        // find the previous node
        while (p != NULL && p->next != NULL && strcmp(p->next->name, name) != 0) {
            p = p->next;
        }

        if (p->next != NULL) {
            d = p->next;
            p->next = d->next;
            free(d);
        } else {
            exit_with_error("no bookmark found");
        }
    }
}

// Saves the bookmarks list in the bookmarks file.
void save_bookmarks() {
    FILE *f;
    bookmark *p = bookmarks;

    if ((f = fopen(get_bookmarks_filepath(), "w")) != NULL) {
        while (p != NULL) {
            fprintf(f, "%s\t%s\n", p->name, p->path);
            p = p->next;
        }
    } else {
        exit_with_error("can't save configuration file");
    }
}

// Loads the bookmarks list from the bookmarks file.
void load_bookmarks() {
    FILE *f;
    char *name, *path;
    name = calloc(NAME_SIZE, sizeof(char));
    path = calloc(PATH_SIZE, sizeof(char));

    if ((f = fopen(get_bookmarks_filepath(), "r")) != NULL) {
        while (fscanf(f, "%s\t%s", name, path) != EOF) {
            add_bookmark(name, path);
        }
    }
}

// Prints the bookmarks list.
void print_bookmarks() {
    bookmark *p = bookmarks;
    int i;

    if (p != NULL) {
        printf("Bookmark             Path\n");
        printf("---------------------------------------------------------------------------\n");
        while (p != NULL) {
            printf("%s", p->name);
            for (i = 0; i < (20 - strlen(p->name)); i++) {
                printf(" ");
            }
            printf(" %s\n", p->path);
            p = p->next;
        }
        
    } else {
        printf("No bookmarks saved.\n");
    }
}

// Expands paths that could start with a bookmark (e.g. [bookmark]/sub/path)
char *expand_path(char *path_with_bookmark) {
    const char *subpath = strchr(path_with_bookmark, '/');

    if (subpath != NULL) {
        // the path contains a subpath (e.g. [bookmark]/sub/path)
        if (subpath == path_with_bookmark) {
            // the path is an absolute path (no bookmark, e.g. /absolute/path)
            char *pwb = strdup(path_with_bookmark);
            return pwb;
        } else {
            // the path is composed of a bookmark and a subpath

            // copy the bookmark part in a variable
            char *b;
            b = calloc(subpath - path_with_bookmark, sizeof(char));
            strncpy(b, path_with_bookmark, (int)(subpath - path_with_bookmark));

            // find the corresponding path
            char *b_path;
            b_path = find_bookmark(b)->path;

            // concatenate the bookmark path with the subpath
            char *expanded_path;
            expanded_path = calloc(strlen(b_path) + strlen(subpath), sizeof(char));
            strcpy(expanded_path, b_path);
            strcat(expanded_path, subpath);

            return expanded_path;
        }
    } else {
        // the path is just a bookmark
        return find_bookmark(path_with_bookmark)->path;
    }
}

int main(int argc, char *argv[]) {
    if (argc > 1) {
        if(strcmp(argv[1], "list") == 0) {
            load_bookmarks();
            print_bookmarks();
        } else if(strcmp(argv[1], "help") == 0) {
            print_usage();
        } else if(strcmp(argv[1], "to") == 0 && argc > 2) {
            load_bookmarks();
            
            char *path;
            path = expand_path(argv[2]);
            if (path != NULL) {
                printf("%s", path);
            }
        } else if(strcmp(argv[1], "add") == 0 && argc > 2) {
            load_bookmarks();
            new_bookmark(argv[2], cwd());
            save_bookmarks();
        } else if(strcmp(argv[1], "del") == 0 && argc > 2) {
            load_bookmarks();
            delete_bookmark(argv[2]);
            save_bookmarks();
        } else {
            print_usage();
            exit(1);
        }
    } else { // wrong number of arguments
        print_usage();
        exit(1);
    }

    return 0;
}


jump.rb
#!/usr/bin/env ruby 

# Jump, a bookmarking system for the bash shell.
# Copyright (c) 2010 Giuseppe Capizzi
# mailto: g.capizzi@gmail.com
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

BOOKMARKS_PATH = File.expand_path "~/.jumprc"

def exit_with_error(message)
  puts "*** ERROR: " + message
  exit 1
end

def print_usage
  puts "jump to [bookmark]    Jumps to the directory pointed by [bookmark]"
  puts "jump add [bookmark]   Saves the current directory in [bookmark]"
  puts "jump del [bookmark]   Deletes [bookmark]"
  puts "jump list             Prints the list of all saved bookmarks"
  puts "jump help             Displays this message"
end

# Saves the bookmarks list in the bookmarks file.
def save_bookmarks(bookmarks, bookmarks_path)
  begin
    file = File.open(bookmarks_path, 'w') do |f|
      bookmarks.each do |name, path|
        f.puts name + "\t" + path
      end
    end
  rescue
    exit_with_error "can't save configuration file"
  end
end

# Loads the bookmarks list from the bookmarks file.
def load_bookmarks(bookmarks_path)
  bookmarks = {}
  
  begin
    file = File.open(bookmarks_path, 'r') do |f|
      while line = f.gets  
        name, path = line.split("\t")
        bookmarks[name] = path
      end
    end
    
    return bookmarks
  rescue
    exit_with_error "can't load configuration file"
  end
end

# Prints the bookmarks list.
def print_bookmarks(bookmarks)
  name_col_title = "Bookmark"
  path_col_title = "Path"
  name_col_size = 20
  row_length = 75

  if !bookmarks.empty?
    print name_col_title
    (name_col_size - name_col_title.length).times do print " " end
    puts path_col_title
    puts "-" * row_length

    bookmarks.each do |name, path|
      print name
      (name_col_size - name.length).times do
        print " "
      end
      puts " " + path
    end
  else
    puts "No bookmarks saved"
  end
end

# Expands paths that could start with a bookmark (e.g. [bookmark]/sub/path)
def expand_path(path_with_bookmark, bookmarks)
  if path_with_bookmark.index("/").nil?
    # the path is just a bookmark
    return bookmarks[path_with_bookmark]
  elsif path_with_bookmark.index("/") == 0
    # the path is an absolute path (no bookmark, e.g. /absolute/path)
    return path_with_bookmark
  else
    # the path is composed of a bookmark and a subpath
    name, path = path_with_bookmark.split("/")
    return bookmarks[name] + "/" + path;
  end
end

if __FILE__ == $PROGRAM_NAME
  if ARGV.size > 0
    if ARGV[0] == "list"
      print_bookmarks(load_bookmarks(BOOKMARKS_PATH))
      # p load_bookmarks(BOOKMARKS_PATH).to_s
    elsif ARGV[0] == "help"
      print_usage
    elsif ARGV[0] == "to" && ARGV.size > 1
      path = expand_path(ARGV[1], load_bookmarks(BOOKMARKS_PATH));
      print path.chomp if !path.nil?
    elsif ARGV[0] == "add" && ARGV.size > 1
      bookmarks = load_bookmarks(BOOKMARKS_PATH)
      bookmarks[ARGV[1]] = Dir.pwd
      save_bookmarks(bookmarks, BOOKMARKS_PATH)
    elsif ARGV[0] == "del" && ARGV.size > 1
      bookmarks = load_bookmarks(BOOKMARKS_PATH)
      bookmarks.delete ARGV[1]
      save_bookmarks(bookmarks, BOOKMARKS_PATH)
    else
      print_usage
      exit 1
    end
  else # wrong number of arguments
    print_usage
    exit 1;
  end
end


lib/bookmarks.rb
# This file is part of the jump project
#
# Copyright (C) 2010 Flavio Castelli <flavio@castelli.name>
# Copyright (C) 2010 Giuseppe Capizzi <gcapizzi@gmail.com>
#
# jump is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# jump is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Keep; if not, write to the
# Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

require 'rubygems'
require 'yaml'
require 'terminal-table/import'

class Bookmarks
  BOOKMARKS_PATH = File.expand_path "~/.jumprc"

  def initialize
    # Loads the bookmarks list from the bookmarks file.
    begin
      @bookmarks = YAML::load_file(BOOKMARKS_PATH)
    rescue Errno::ENOENT
      @bookmarks = {}
    rescue
      raise "Can't load configuration file"
    end
  end

  # Checks if +bookmark+ name is valid
  def self.is_valid_name? bookmark
    return (bookmark =~/\A\W/).nil?
  end

  # Saves the bookmarks list in the bookmarks file.
  def save
    begin
      File.open(BOOKMARKS_PATH, 'w') do |file|
        file << YAML::dump(@bookmarks)
      end
    rescue
      raise "Can't save configuration file"
    end
  end

  # Adds +bookmark+ pointing to +path+
  def add path, bookmark
    @bookmarks[bookmark] = path
  end

  # Deletes +bookmark+
  def del bookmark
    if @bookmarks.has_key? bookmark
      @bookmarks.delete bookmark
      true
    else
      false
    end
  end

  # Prints the bookmarks list.
  def to_s
    if @bookmarks.empty?
      "No bookmarks saved"
    else
      bookmarks_table = table do |t|
        t.style = { :border_y => '', :border_i => '' }
        t.headings = "Bookmark", "Path"
        @bookmarks.keys.sort.each do |bookmark|
          t << [bookmark, simplify_path(@bookmarks[bookmark])]
        end
      end
      bookmarks_table.to_s
    end
  end

  def sorted_bookmarks()  sorted_list @bookmarks.keys  end
  def sorted_list(terms)  terms.sort.join ' '  end

  # Provide a list of completion options, starting with given prefix
  def complete prefix
    # Special cases:
    #   - nothing is provided: return all the bookmarks
    #   - absolute path: don't complete
    return sorted_bookmarks if prefix.nil? || prefix.empty?
    return prefix if prefix.start_with? File::SEPARATOR

    bookmark, path = prefix.split File::SEPARATOR, 2 # File.split only does path/basename

    completions = [  ]
    if path.nil?
      # still in 1st element, could match several bookmarks
      completions += @bookmarks.keys.find_all { |b| b.start_with? prefix }
    elsif @bookmarks.has_key?(bookmark)
      # bookmark known, complete further
      completions += Dir.chdir(@bookmarks[bookmark]) do
        Dir.glob(["#{path}*"]) \
          .select { |f| File.directory? f } \
          .collect { |f| File.join bookmark, f }
      end
    end
    completions.map! { |d| d + File::SEPARATOR }
    completions << prefix if completions.empty?
    sorted_list completions
  end

  # Simplifies given path by replacing the user's homedir with ~
  def simplify_path(path)
    path.gsub /^#{File.expand_path '~'}/, '~'
  end

  # Expands paths that could start with a bookmark (e.g. [bookmark]/sub/path)
  def expand_path(path_with_bookmark)
    if path_with_bookmark.index("/").nil?
      # the path is just a bookmark
      return @bookmarks[path_with_bookmark]
    elsif path_with_bookmark.index("/") == 0
      # the path is an absolute path (no bookmark, e.g. /absolute/path)
      return path_with_bookmark
    else
      # the path is composed of a bookmark and a subpath
      name = path_with_bookmark[0, path_with_bookmark.index('/')]
      path = path_with_bookmark[path_with_bookmark.index('/')+1,
                                path_with_bookmark.size]
      return @bookmarks[name] + "/" + path;
    end
  end
end
